docker-build-system/zombodb-build-alpine-3.12/Dockerfile

```diff
@@ -0,0 +1,75 @@
+#
+# Copyright 2018-2020 ZomboDB, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+FROM rust:1.47-alpine3.12 as zombodb-extension
+
+LABEL maintainer="ZomboDB, LLC (zombodb@gmail.com)"
+
+ARG USER=docker
+ARG UID=1000
+ARG GID=1000
+RUN adduser --uid=${UID} -D ${USER}
+
+RUN apk add --no-cache \
+	musl-dev \
+	openssl-dev \
+	make \
+	bison \
+	coreutils \
+	dpkg-dev dpkg \
+	flex \
+	gcc \
+	libc-dev \
+	libedit-dev \
+	libxml2-dev \
+	libxslt-dev \
+	linux-headers \
+	make \
+	openssl-dev \
+	perl-utils \
+	perl-ipc-run \
+	util-linux-dev \
+	zlib-dev \
+	icu-dev
+
+USER ${UID}:${GID}
+WORKDIR /home/${USER}
+
+RUN cargo install cargo-pgx
+COPY build_pg.sh build_pg.sh
+
+ENV PG_10=10.15
+RUN wget https://ftp.postgresql.org/pub/source/v${PG_10}/postgresql-${PG_10}.tar.gz -qO /home/${USER}/postgresql-10.tar.gz
+RUN sh ./build_pg.sh 10
+
+ENV PG_11=11.10
+RUN wget https://ftp.postgresql.org/pub/source/v${PG_11}/postgresql-${PG_11}.tar.gz -qO /home/${USER}/postgresql-11.tar.gz
+RUN sh ./build_pg.sh 11
+
+ENV PG_12=12.5
+RUN wget https://ftp.postgresql.org/pub/source/v${PG_12}/postgresql-${PG_12}.tar.gz -qO /home/${USER}/postgresql-12.tar.gz
+RUN sh ./build_pg.sh 12
+
+ENV PG_13=13.1
```

Comment:
these envvars are going to be untenable for me to keep up with.  Whenever I upgrade `pgx` and then `zombodb` to compile against whatever latest version of PG, then I'll also have to come and edit this Dockerfile.  That's a workflow killer.

Any ideas on how we can fix this?  What is Alpine and why doesn't it have packages for PG10/11/12/13?

Reply:
Alpine is a linux distribution popular to build docker image because it’s very small. The base image is only ~ 5mb

It uses apk as its package manager and apparently they don’t keep old version of their package for too long : https://gitlab.alpinelinux.org/alpine/abuild/-/issues/9996#note_87135

(and they don’t have a postgresql 13 package yet)

I’ll try to think of something to avoid having to update env vars here

Reply:
I can find the minor version of a branch like that : 

`curl -s http://ftp.postgresql.org/pub/source/ | grep v10\\. | sed 's/.*href="v10\.\([0-9]*\)\/.*/\1/' | sort -nr | head -n1`

and use that to fetch and build the latest version of postgres but I don’t like it very much. it looks like something that could break...

---

I could put those env var somewhere else (in /docker-build-system/postgresql_versions.env) but I don’t know if it’s much better for you and it would be odd because it’s not used in all Dockerfiles. maybe be naming it /docker-build-system/postgresql_from_source_versions.env ?